omod/src/main/java/org/openmrs/module/auditlogweb/web/controller/ViewAuditController.java

```diff
@@ -0,0 +1,144 @@
+            } catch (org.hibernate.ObjectNotFoundException ex) {
+                model.addAttribute("errorMessage", "Audit entity not found for this revision.");
+                return new ModelAndView(VIEW, model);
+            }
+
+            if (auditId - 1 > 0) {
+                try {
+                    oldEntity = auditService.getRevisionById(clazz, entityId, auditId - 1);
+                } catch (org.hibernate.ObjectNotFoundException ex) {
+                    oldEntity = null;
+                }
+            }
+            List<AuditFieldDiff> diffs = computeFieldDiffs(clazz, oldEntity, currentEntity);
+            model.addAttribute("diffs", diffs);
+            return new ModelAndView(VIEW, model);
+
+        } catch (IllegalArgumentException e) {
+            if (e.getCause() instanceof QueryException) {
+                model.addAttribute("errorMessage", "Some referenced data could not be loaded. Please contact your administrator.");
+                return new ModelAndView(VIEW, model);
+            } else {
+                throw e;
+            }
+        } catch (Exception e) {
+            model.addAttribute("errorMessage", "Error loading audit data: " + e.getMessage());
+            e.printStackTrace();
+            return new ModelAndView(VIEW, model);
+        }
+    }
+
+    private List<AuditFieldDiff> computeFieldDiffs(Class<?> clazz, Object oldEntity, Object currentEntity) {
+        List<AuditFieldDiff> diffs = new ArrayList<>();
+
+        if (currentEntity == null) {
+            return diffs;
+        }
+
+        Field[] fields = clazz.getDeclaredFields();
+        for (Field field : fields) {
+            if (java.lang.reflect.Modifier.isStatic(field.getModifiers()) || field.isSynthetic()) {
+                continue;
+            }
+
+            field.setAccessible(true);
+            Object currentValue = null;
+            Object oldValue = null;
+
+            try {
+                currentValue = field.get(currentEntity);
+            } catch (IllegalAccessException ignored) {
+            }
+
+            try {
+                if (oldEntity != null) {
+                    oldValue = field.get(oldEntity);
+                }
+            } catch (IllegalAccessException ignored) {
+            }
+
+            String oldStr = safeToString(oldValue);
+            String currStr = safeToString(currentValue);
+
+            boolean isDifferent = !Objects.equals(oldStr, currStr);
+
+            diffs.add(new AuditFieldDiff(field.getName(), oldStr, currStr, isDifferent));
+        }
+
+        return diffs;
+    }
+
+    private String safeToString(Object obj) {
+        if (obj == null) return "null";
+        try {
+            return obj.toString();
+        } catch (org.hibernate.ObjectNotFoundException | org.hibernate.LazyInitializationException e) {
+            return "Data not found";
+        } catch (Exception e) {
+            return "Data not found";
+        }
+    }
+
+}
```

Comment:
Could we move this to a different location? This does not feel like it should belong in a controller class.